File: 2038-nearest-exit-from-entrance-in-maze/nearest-exit-from-entrance-in-maze.py
from collections import deque

class Solution:
    def nearestExit(self, maze: List[List[str]], entrance: List[int]) -> int:
        rows, cols = len(maze), len(maze[0])
        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        entrance_row, entrance_col = entrance
        
        queue = deque([(entrance_row, entrance_col, 0)])  # (row, col, steps)
        maze[entrance_row][entrance_col] = '#'  # Mark entrance as visited
        
        while queue:
            row, col, steps = queue.popleft()
            
            if (row != entrance_row or col != entrance_col) and (row == 0 or row == rows - 1 or col == 0 or col == cols - 1):
                return steps
            
            for dr, dc in directions:
                r, c = row + dr, col + dc
                if 0 <= r < rows and 0 <= c < cols and maze[r][c] == '.':
                    maze[r][c] = '#'  # Mark cell as visited
                    queue.append((r, c, steps + 1))
        
        return -1

File: 2038-nearest-exit-from-entrance-in-maze/nearest-exit-from-entrance-in-maze.cpp
#include <vector>
#include <queue>

using namespace std;

class Solution {
public:
    int nearestExit(vector<vector<char>>& maze, vector<int>& entrance) {
        int rows = maze.size();
        int cols = maze[0].size();
        vector<pair<int, int>> directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        int entrance_row = entrance[0];
        int entrance_col = entrance[1];
        
        queue<tuple<int, int, int>> q;  // (row, col, steps)
        q.push({entrance_row, entrance_col, 0});
        maze[entrance_row][entrance_col] = '#';  // Mark entrance as visited
        
        while (!q.empty()) {
            int row, col, steps;
            tie(row, col, steps) = q.front();
            q.pop();
            
            if ((row != entrance_row || col != entrance_col) && (row == 0 || row == rows - 1 || col == 0 || col == cols - 1)) {
                return steps;
            }
            
            for (auto& [dr, dc] : directions) {
                int r = row + dr;
                int c = col + dc;
                if (r >= 0 && r < rows && c >= 0 && c < cols && maze[r][c] == '.') {
                    maze[r][c] = '#';  // Mark cell as visited
                    q.push({r, c, steps + 1});
                }
            }
        }
        
        return -1;
    }
};
